Fix ascent

SteffenBauer opened the pull request:
This pull request should solve similiar problems like in bricks:

a) Set `n = -1` when `i == w*h` in `new_game_desc`, as n is not needed for this condition, to avoid invalid access.
b) `n` can get negative in `game_redraw`, causing negative index access to `*positions`. Add extra check for n to be non-negative.

SteffenBauer committed:
Bricks: a) Fix invalid grid accesses b) Fix solve operation c) Switch to new MOVE_ return codes

SteffenBauer committed:
Fix two illegal array accesses in ascent puzzle

x-sheep commented:
Could you rebase this onto the latest version of `master`? This PR currently contains the commit for Bricks as well.

SteffenBauer commented:
Yes, already saw that, I plan to do this evening.

SteffenBauer committed:
Merge branch 'x-sheep:master' into master

SteffenBauer committed:
Merge branch 'master' into fix_ascent

SteffenBauer commented:
Should be mergeable now

x-sheep commented on ascent.c:
```diff
@@ -3782,7 +3782,7 @@ static void game_redraw(drawing *dr, game_drawstate *ds,
 					GET_BIT(state->immutable, i) ? COL_IMMUTABLE : 
 					state->grid[i] == NUMBER_EMPTY && ui->typing_cell != i ? COL_LOWLIGHT :
-					n <= state->last && positions[n] == CELL_MULTIPLE && ui->typing_cell != i ? COL_ERROR :
+					n >= 0 && n <= state->last && positions[n] == CELL_MULTIPLE && ui->typing_cell != i ? COL_ERROR :
```
Is this needed? Line 3777 already checks if `n >= 0`

x-sheep commented on ascent.c:
```diff
@@ -1743,7 +1743,7 @@ static char *new_game_desc(const game_params *params, random_state *rs,
 	for(i = 0; i <= w*h; i++)
 	{
-		n = grid[i];
+		n = (i == w*h) ? -1 : grid[i];
```
I prefer using one of the `NUMBER_*` constants instead of typing `-1`

SteffenBauer committed:
Use constant; remove unneeded check

SteffenBauer commented:
> Is this needed? Line 3777 already checks if n >= 0

Actually not ... I removed the check, made a fresh compile, and again run with valgrind. No illegal accesses. No clue why exactly I added that check in the first place, and what valgrind complained about then.

First change request, using one of the defined constants, done. Pull request very short now.
